src/Button.spec.tsx

```diff
@@ -0,0 +1,200 @@
+// transpiled files that we can't import.
+import { IconShip2 } from "../icons/IconShip2.tsx";
+
+describe("Button", () => {
+  afterEach(cleanup);
+
+  (["light", "dark"] as const).forEach(theme => {
+    describe(`for ${theme} theme`, () => {
+      describe("should render default button", () => {
+        it("with default size", () => {
+          const { getByText } = render(<Button theme={theme}>submit</Button>);
+          expect(getByText("submit")).toBeInTheDocument();
+        });
+
+        it("with small size", () => {
+          const { getByText } = render(
+            <Button size="small" theme={theme}>
+              submit
+            </Button>
+          );
+          expect(getByText("submit")).toBeInTheDocument();
+        });
+
+        it("with large size", () => {
+          const { getByText } = render(
+            <Button size="large" theme={theme}>
+              submit
+            </Button>
+          );
+          expect(getByText("submit")).toBeInTheDocument();
+        });
+
+        it("with an icon", () => {
+          const { container, getByText } = render(
+            <Button icon={<IconShip2 />} theme={theme}>
+              submit
+            </Button>
+          );
+          expect(getByText("submit")).toBeInTheDocument();
+          expect(container.querySelector("svg")).toBeInTheDocument();
+        });
+
+        it("disabled", () => {
+          const { getByText } = render(
+            <Button disabled={true} theme={theme}>
+              submit
+            </Button>
+          );
+          expect(getByText("submit")).toBeInTheDocument();
+        });
+      });
+
+      describe("should render flat button", () => {
+        it("with default color", () => {
+          const { getByText } = render(
+            <Button feel="flat" theme={theme}>
+              submit
+            </Button>
+          );
+          expect(getByText("submit")).toBeInTheDocument();
+        });
+
+        it("with custom color", () => {
+          const { getByText } = render(
+            <Button feel="flat" theme={theme} color={colors.blue.base}>
+              submit
+            </Button>
+          );
+          expect(getByText("submit")).toBeInTheDocument();
+        });
+      });
+
+      describe("should render icon only button", () => {
+        it("with default size", () => {
+          const { getByTestId } = render(
+            <Button theme={theme} icon={<IconShip2 />} data-testid="button" />
+          );
+          expect(getByTestId("button")).toBeInTheDocument();
+        });
+
+        it("with small size", () => {
+          const { getByTestId } = render(
+            <Button
+              theme={theme}
+              icon={<IconShip2 />}
+              data-testid="button"
+              size="small"
+            />
+          );
+          expect(getByTestId("button")).toBeInTheDocument();
+        });
+
+        it("with large size", () => {
```

Comment:
With these tests, we aren't checking the size of a large button, or the appearance of a flat button, for example. Is that something we can do with jest? Is that something we want to do here, or does chromatic cover that?

Reply:
This is a great question. I need to document exactly what the expectations are for tests or no one will know what to write on their own.

These tests are intended to test _functionality_. That means when we pass stuff, those things are rendered as expected. All behaviors (like clicking the mouse or pressing Escape) should be tested.

These tests are incapable of testing visual appearance because they are never painted. Jest uses jsdom to render the dom; but it doesn't actually lay that out or paint it; it just produces a dom that we can analyze. We'll need to use Chromatic to visually test things.

One more note: these tests are expected to be fast. Visual testing is expected to be slow. I would expect that we can develop with TDD (write tests before writing code) and then watching jests's output as we work to make sure we haven't broken anything. I expect Chromatic to run in CI, not on local workstations because it's slow.